### PR TITLE
[CS2113-T11-1] Wellness360

### PR DESCRIPTION
**Pull request by genexus85**
Wellness360 is a wellness app that is meant for stressed Engineering Students who prefer CLI over GUI and want to keep track of their overall wellbeing. Our app offers comprehensive tracking and management tools, providing personalized support to alleviate stress and enhance overall well-being.

### CODE REVIEW

File: docs/diagrams/storage/StorageSequenceDiagramSleep.png
Comment (BryanCastorius) [inline docs/diagrams/storage/StorageSequenceDiagramSleep.png:1]: Shouldn't all the activation bars be enclosed (e.g. some activation bars, top and bottom parts are not closed by a line)?

Comment (BryanCastorius) [inline docs/diagrams/storage/StorageSequenceDiagramSleep.png:1]: Aren't all classes supposed to start with a semicolon (e.g. :SleepTrackerStorage)?

File: docs/diagrams/focus/FocusSequenceDiagram.png
Comment (BryanCastorius) [inline docs/diagrams/focus/FocusSequenceDiagram.png:1]: For the activation bar underneath the :PlaceholderFocusCommand class, there's a deletion (X), however after the (X), is the lifeline supposed to still be continued?

File: docs/diagrams/sleep/Sleep.puml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+SleepCommand --> SleepTracker : uses >
+
+
+SleepCycleList ..> Ui : uses >
+SleepTracker *--> "1" SleepCycleList: contains >
+SleepTracker ..> SleepTrackerStorage : uses >
+
+SleepCycleList *-- "*" SleepCycle : contains
+
+@enduml
```
Comment (tinaliu27) [inline docs/diagrams/sleep/Sleep.puml:95]: For this class diagram, it was a bit confusing trying to following the creates inheritance from SleepCommandParser and Command. Perhaps try stretching out the diagram to make the lines more clear?

Comment (tinaliu27) [reply docs/diagrams/sleep/Sleep.puml:95]: Otherwise great job with the diagram! Very detailed :)

File: docs/diagrams/storage/StorageSequenceSleep.png
Comment (BryanCastorius) [inline docs/diagrams/storage/StorageSequenceSleep.png:1]: When invoking a method, we do not need to write in "calls..." right?

File: docs/diagrams/focus/FocusStateDiagram.puml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+StateReady --> StateTimerRunning: start
+StateTimerRunning -up-> StateReady: stop
+StateTimerRunning -up-> StatePaused: pause
+StatePaused --> StateTimerRunning: resume
+StatePaused --> StateReady: stop
+StatePaused --> StatePaused: check
+StateTimerRunning --> StateTimerRunning: check
+
+
+@enduml
```
Comment (tinaliu27) [inline docs/diagrams/focus/FocusStateDiagram.puml:17]: Great job with the diagrams! Just make sure that when you are using <<class>> you wanna use a lower case!

**Review by BryanCastorius — COMMENTED** [review]
Overall, a complete class and sequence diagrams have been provided with comprehensive explanations. Just a few minor convention error here and there on the sequence diagram.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,38 +1,1282 @@
...
+states. For example, the command `focus pause` cannot be used if the timer hasn't started. Another example would be 
+the `focus switch` command to switch between the timer could not be used if the current timer mode is running. To aid 
+the understanding of the logic, we will use state transition diagram.
+
+![FocusStateDiagram.png](diagrams/focus/FocusStateDiagram.png)
+* The black circle in the diagram represents the starting point of focus timer. 
+* The labels of the arrows represents the commands.
+
+#### Focus Sequence Diagram
+![FocusSequenceDiagram.png](diagrams/focus/FocusSequenceDiagram.png)
```
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:797]: For the sequence diagram step 7, maybe it be consistent with other sequence diagrams step 7?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,38 +1,1282 @@
...
+    * Encapsulation: Private access modifiers and encapsulated methods ensure data integrity. Methods like addSleepCycle
+    and deleteSleepCycle in SleepCycleList encapsulate the manipulation of the reflection list, ensuring data integrity
+    and promoting a clear interface for interacting with the list.
+    * Exception Handling: Extensive coverage of exceptions in sleepCommand classes to ensure all errors are handled 
+    properly
+
+#### Implementation
+
+##### Class Diagram
+![SleepDiagram.png](diagrams/sleep/SleepDiagram.png)
```
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:616]: Is the this class diagram too messy?

File: docs/DeveloperGuide.md
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:1]: Overall good job on the very descriptive developer guide. But maybe consider that it might be too wordy?

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,1282 @@
...
+    - _Exception Handling_: Exception Handling prevents the code from reaching an unknown or unpredictable state, which could break the program.
+    - _Design Pattern_: To Be Continued
+    - _Code Readability_: Proper coding convention, Java Docs and comments were added for clarity so that other developers can more easily review our code.
+<!-- Modularity, Inheritance, Encapsulation, Exception Handling, Design Pattern, readability, etc --> 
+
+#### Implementation
+#### Class Diagram
+<!-- Insert image and description of each class, with its overview, attributes,
+methods, dependencies and UML Notes -->
+![FitnessClassDiagram](./diagrams/fitness/FitnessClassDiagram.png)
```
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:838]: Is this too small for viewing?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,1282 @@
...
+Note that certain details described below have been omitted from Class Diagram for simplicity and to improve readability.
```
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:840]: Is there suppose to be a newline after the diagram? There is minor formatting issue

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,1282 @@
...
+- #### Developer Design Considerations
+  - Static Attributes: All methods are static, allowing their usage anywhere within the codebase.
+  - SRP: Classes adhere to the Single Responsibility Principle. For example, the `Ui` class is solely responsible for managing printing-related operations.
+  - Readability and Maintainability: Descriptive naming, comments, and documentation for clarity. Since printing is a highly used task, it prevents highly-repetitive code as well.
+
+#### Implementation
+
+#### Class Diagram
+<!-- Insert Class Diagram here -->
+![UiClassDiagram.png](diagrams/Ui/UiClassDiagram.png)
```
Comment (cheeseong2001) [inline docs/DeveloperGuide.md:139]: Is the coloured "C" icon supposed to be present?

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,1282 @@
...
+3. Read the developer guide to understand the design.
 
 ## Design & implementation
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### Architecture
+
+### UI component
+#### Description
+`Ui` component is in charge of reading in user input and printing output. </br> </br>
```
Comment (rismm) [inline docs/DeveloperGuide.md]: It appears that the line break in this line is in the wrong syntax? Maybe use `<br>` instead or remove the line break altogether.

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,1282 @@
...
+  writing/reading of files, while HabitTrackerStorage/SleepTrackerStorage deals with the processing of data for storage 
+  and reading for their individual components
+  * Readability and Maintainability: Ensure there are no magic numbers used. For example, use of constants for 
+  information position in the text file.
+  * Exception Handling: Graceful error handling with meaningful messages.
+
+#### Implementation
+
+##### Class Diagram
+![StorageClassDiagram.png](diagrams/storage/StorageClassDiagram.png)
```
Comment (rismm) [inline docs/DeveloperGuide.md:268]: Do consider adding `:void` to void methods in the class diagram for better clarity

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,1282 @@
...
+  * Exception Handling: Graceful error handling with meaningful messages.
+  * Command Design Pattern: Encapsulation of actions as objects for decoupling.
+  * Dependency Injection: Dependencies injected via constructors for loose coupling.
+  * SRP: Classes adhere to the Single Responsibility Principle. For example, the ReflectionManager class is responsible for managing reflection-related operations, while the ReflectionCommandParser class handles parsing and determining reflection commands.
+  * Readability and Maintainability: Descriptive naming, comments, and documentation for clarity.
+
+#### Implementation
+
+##### Class Diagram
+![ReflectionClassDiagram.png](diagrams/reflection/ReflectionClassDiagram.png)
```
Comment (rismm) [inline docs/DeveloperGuide.md:362]: Might want to simplify or make the diagram's font larger. The class diagram is a little hard to see and is quite convoluted

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,1282 @@
...
+      * `FAVOURITE_QUESTIONS_FILE_PATH`: Path to the file storing favorite reflection questions. 
+      * `fiveRandomQuestions`: ArrayList holding five random reflection questions. 
+      * `questionBank`: Instance of `ReflectionQuestionBank` managing reflection questions.
+      * `favoriteReflectionsList`: Instance of `FavoriteReflectionsList` managing favorite reflection questions.
+    * Methods:
+      * `printFiveRandomQuestions()`: Prints five random reflection questions.
+      * `saveReflectionQuestion(int reflectionId)`: Saves a reflection question to favorites based on ID.
+      * `unsaveReflectionQuestion(int reflectionId)`: Removes a reflection question from favorites based on ID.
+      * `printFavourites()`: Prints the list of favorite reflection questions.
+      * `printHelpMenu()`: Prints the help menu with available reflection commands.
```
Comment (rismm) [inline docs/DeveloperGuide.md:379]: Maybe avoid explaining methods where the method names are quite self-explanatory to avoid over explanation.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,1282 @@
...
+![StorageSequenceDiagramSleep.png](diagrams/storage/StorageSequenceDiagramSleep.png)
+
+* Note: HabitTrackerStorage works in a similar way as SleepTrackerStorage
+
+When `Main` starts, `SleepTracker` object is created. `SleepTrackerStorage loadSleepListFromFile()` method is called to 
+Begin loading sleep cycles from text file. `SleepCycleList` is returned to `SleepTracker` class and Sleep Tracker class 
+is successfully instantiated.
+
+##### Saving Data Sequence Diagram
+![StorageSequenceSleep.png](diagrams/storage/StorageSequenceSleep.png)
```
Comment (rismm) [inline docs/DeveloperGuide.md:324]: Consider hiding the footbox of classes at the bottom of the sequence diagram

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,1282 @@
...
+    * Modularity: Encapsulate related functionalities within classes to promote re-usability and maintainability.
+    * Exception Handling: Use of custom exceptions to differentiate between various types of errors, and to handle them appropriately with error messages
+    * Data Encapsulation: Control the access to a class's internal attributes, accessible only through getters and setters
+    * Interface Segregation: Segregation of command interface to represent different command types for specific use cases.
+    * Readability and Maintainability: Descriptive naming, robust documentation for code clarity.
+
+#### Implementation
+
+##### Class Diagram
+![HabitClassDiagram.png](diagrams/habit/HabitClassDiagram.png)
```
Comment (rismm) [inline docs/DeveloperGuide.md:500]: Could consider excluding some methods and parameters to simplify the diagram as it looks too bulky.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,1282 @@
...
+        * `HabitTracker()`: Constructs a HabitTracker object. 
+        * `getNumberOfHabits()`: Returns the number of habits in the habitList.
+        * `addHabit(Habit newHabit)`: Adds a new Habit object into the habitList.
+        * `listHabits()`: Prints a list of all habits in habitList .
+        * `isValidHabitID(int habitID)`: Check if a habit ID is valid by comparing with the size of habitList.
+        * `updateHabitCount(int habitID, String updatedCount)`: Update the habit count of a habit.
+        * `deleteHabit(int habitID)`: Delete a habit from habitList.
+        * `setPriorityLevel(int habitID, String priority)`: Set the priority of a habit.
+        * `sortHabits()`: Sort the habits in habitList according to the habits' priorities.
+        * `clearHabitList()`: Delete all habits from habitList.
```
Comment (rismm) [inline docs/DeveloperGuide.md:517]: Could consider leaving out listing methods where the method names are quite self-explanatory

**Review by rismm — COMMENTED** [review]
Overall good job on the DG! Do consider simplifying the DG as it seems a little too overwhelming.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (ParthGandhiNUS) [inline docs/DeveloperGuide.md:616]: While this diagram is very comprehensive, perhaps can consider breaking it up into more manageable smaller diagrams. Having so many lines which are overlapping can be confusing to developers.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,1282 @@
...
+* Our app offers comprehensive tracking and management tools, providing personalized support to alleviate stress and enhance overall well-being.
+* It can help you take control of your mental and physical health effortlessly, so you can focus on your studies with peace of mind.
 
 ## User Stories
 
-|Version| As a ... | I want to ... | So that I can ...|
-|--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+| Version | As a ... | I want to ...                                                                         | So that I can ...                                                   |
```
Comment (ParthGandhiNUS) [inline docs/DeveloperGuide.md:1033]: Comprehensive User stories being used. Good Job!! Warms the cockles in my heart!

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (ParthGandhiNUS) [inline docs/DeveloperGuide.md:268]: Class diagram looks neat and comprehensive! Good Job!

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (hafizuddin-a) [inline docs/DeveloperGuide.md:838]: Detailed diagram. Might want to consider resizing and insert newline after the diagram

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,1282 @@
...
+    * `SetPriorityCommand`: Set priority level of habit
+        * Command format: `habit set /id <habit_ID> /priority <priority_level>`
+    * `SortHabitsCommand`: Sort habit list according to priority level.
+        * Command format: `habit sort`
+    * `HabitHelpCommand`: Display a help menu of the habit-tracker commands
+        * Command format: `habit help`
+
+
+#### Sequence Diagram
+![HabitSequenceDiagram.png](diagrams/habit/HabitSequenceDiagram.png)
```
Comment (hafizuddin-a) [inline docs/DeveloperGuide.md:568]: Perhaps the words on this diagram may be hard to read

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,1282 @@
...
+  - Attributes:
+    - `SEP`: A private static string constant that stores a separation line.
+  - Methods:
+    - `greetUser()`: A public static method that prints a welcome message when Wellness360 is started.
+    - `promptUserInput()`: A public static method that prints a "You: " message before the input cursor, to signal the input of
+    - `sayGoodbye()`: A public static method that prints a farewell message when Wellness360 is terminated by the user.
+    - `printMessagewithSepNewLine(String message)`: A public static method used to print a message with a separation line.
+    - `printMessageWithoutNewLine(String message)`: A public static method used to print a message without a separation line.
+    - `printList(ArrayList list, String message)`: A public static method to print a list of type `ArrayList`, along with a message and a separation line.
+### Command parser component
```
Comment (hafizuddin-a) [inline docs/DeveloperGuide.md:153]: Perhaps accidental empty section

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (hafizuddin-a) [inline docs/DeveloperGuide.md:268]: Is the coloured C supposed to be present

**Review by hafizuddin-a — COMMENTED** [review]
Overall very detailed DG. Good effort
